src/main/scala/is/hail/table/Table.scala

```diff
@@ -38,16 +39,53 @@ object SortColumn {
 
 case class SortColumn(column: String, sortOrder: SortOrder)
 
-case class TableMetadata(
+object TableMetadata {
+  def read(hc: HailContext, path: String): TableMetadata = {
+    val metadataFile = path + "/metadata.json.gz"
+    val jv = hc.hadoopConf.readFile(metadataFile) { in => parse(in) }
+
+    val fileVersion = jv \ "file_version" match {
+      case JInt(rep) => SemanticVersion(rep.toInt)
+      case _ =>
+        fatal(s"metadata does not contain file version: $metadataFile")
+    }
+
+    if (!FileFormat.version.supports(fileVersion))
+      fatal(s"incompatible file format when reading: $path\n  supported version: ${ FileFormat.version }, found $fileVersion")
+
+    val jmetadata = jv.extract[JSONTableMetadata]
```

Comment:
same thing here, we could avoid the boilerplate and the JSONTableMetadata (assuming you're willing to live with underscores in scala or camelcase in JSON).